File: server/routes/friends.js
const express = require('express');
const { db, pair } = require('../db');
const { authMiddleware } = require('../auth');
const router = express.Router();

router.use(authMiddleware);

router.get('/', (req, res) => {
  const me = req.user.uuid;
  const rows = db.prepare(`
    SELECT f.*, 
           ua.username AS username_a, ua.status AS status_a, ua.last_seen AS seen_a,
           ub.username AS username_b, ub.status AS status_b, ub.last_seen AS seen_b
    FROM friendships f
    JOIN users ua ON ua.uuid = f.user_a
    JOIN users ub ON ub.uuid = f.user_b
    WHERE (f.user_a = ? OR f.user_b = ?) AND f.status = 'accepted'
  `).all(me, me);

  const friends = rows.map(r => {
    const isA = r.user_a === me;
    return {
      uuid:     isA ? r.user_b     : r.user_a,
      name:     isA ? r.username_b : r.username_a,
      status:   isA ? r.status_b   : r.status_a,
      lastSeen: isA ? r.seen_b     : r.seen_a,
      since:    r.created_at
    };
  });

  res.json({ friends });
});

router.get('/requests', (req, res) => {
  const me = req.user.uuid;
  
  const incoming = db.prepare(`
    SELECT f.requested_by AS uuid, u.username, f.created_at
    FROM friendships f
    JOIN users u ON u.uuid = f.requested_by
    WHERE (f.user_a = ? OR f.user_b = ?)
      AND f.status = 'pending'
      AND f.requested_by != ?
  `).all(me, me, me);

  const outgoing = db.prepare(`
    SELECT 
      CASE WHEN f.requested_by = ? THEN f.user_b ELSE f.user_a END AS uuid,
      CASE WHEN f.requested_by = ? THEN ub.username ELSE ua.username END AS username,
      f.created_at
    FROM friendships f
    JOIN users ua ON ua.uuid = f.user_a
    JOIN users ub ON ub.uuid = f.user_b
    WHERE (f.user_a = ? OR f.user_b = ?)
      AND f.status = 'pending'
      AND f.requested_by = ?
  `).all(me, me, me, me, me);

  res.json({ incoming, outgoing });
});

router.post('/add', (req, res) => {
  const me = req.user.uuid;
  const { username } = req.body || {};
  if (!username || typeof username !== 'string' || username.length > 16) {
    return res.status(400).json({ error: 'invalid_username' });
  }

  const target = db.prepare('SELECT uuid, username FROM users WHERE username = ? COLLATE NOCASE').get(username);
  if (!target) return res.status(404).json({ error: 'user_not_found', hint: 'L\'utilisateur doit s\'être connecté au moins une fois' });
  if (target.uuid === me) return res.status(400).json({ error: 'cannot_add_self' });

  const [a, b] = pair(me, target.uuid);
  const existing = db.prepare('SELECT * FROM friendships WHERE user_a = ? AND user_b = ?').get(a, b);

  const ws = req.app.get('ws');

  if (existing) {
    if (existing.status === 'accepted') return res.status(409).json({ error: 'already_friends' });
    if (existing.status === 'blocked')  return res.status(403).json({ error: 'blocked' });
    if (existing.requested_by === me)   return res.status(409).json({ error: 'already_requested' });

    db.prepare('UPDATE friendships SET status = ? WHERE id = ?').run('accepted', existing.id);

    if (ws) {
      ws.notify(target.uuid, { type: 'friend_accepted', by: req.user });

      if (ws.isOnline(me)) {
        const meRow = db.prepare('SELECT username, status FROM users WHERE uuid = ?').get(me);
        ws.notify(target.uuid, { type: 'friend_status', uuid: me, name: meRow.username, status: meRow.status, lastSeen: Date.now() });
      }
      if (ws.isOnline(target.uuid)) {
        const otherRow = db.prepare('SELECT username, status FROM users WHERE uuid = ?').get(target.uuid);
        ws.notify(me, { type: 'friend_status', uuid: target.uuid, name: otherRow.username, status: otherRow.status, lastSeen: Date.now() });
      }
    }

    return res.json({ ok: true, status: 'accepted' });
  }

  db.prepare(`
    INSERT INTO friendships (user_a, user_b, status, requested_by, created_at)
    VALUES (?, ?, 'pending', ?, ?)
  `).run(a, b, me, Date.now());

  ws?.notify(target.uuid, { type: 'friend_request', from: req.user });

  res.json({ ok: true, status: 'pending' });
});


router.post('/accept', (req, res) => {
  const me = req.user.uuid;
  const { uuid } = req.body || {};
  if (!uuid) return res.status(400).json({ error: 'missing_uuid' });

  const [a, b] = pair(me, uuid);
  const f = db.prepare('SELECT * FROM friendships WHERE user_a = ? AND user_b = ?').get(a, b);
  if (!f || f.status !== 'pending') return res.status(404).json({ error: 'no_request' });
  if (f.requested_by === me) return res.status(400).json({ error: 'cannot_accept_own' });

  db.prepare('UPDATE friendships SET status = ? WHERE id = ?').run('accepted', f.id);

  const ws = req.app.get('ws');
  ws?.notify(uuid, { type: 'friend_accepted', by: req.user });

  if (ws) {
    if (ws.isOnline(me)) {
      const meRow = db.prepare('SELECT username, status FROM users WHERE uuid = ?').get(me);
      ws.notify(uuid, { type: 'friend_status', uuid: me, name: meRow.username, status: meRow.status, lastSeen: Date.now() });
    }
    if (ws.isOnline(uuid)) {
      const otherRow = db.prepare('SELECT username, status FROM users WHERE uuid = ?').get(uuid);
      ws.notify(me, { type: 'friend_status', uuid, name: otherRow.username, status: otherRow.status, lastSeen: Date.now() });
    }
  }

  res.json({ ok: true });
});

router.delete('/:uuid', (req, res) => {
  const me = req.user.uuid;
  const other = req.params.uuid;
  const [a, b] = pair(me, other);
  
  const existing = db.prepare('SELECT * FROM friendships WHERE user_a = ? AND user_b = ?').get(a, b);
  if (!existing) return res.status(404).json({ error: 'not_found' });
  
  db.prepare('DELETE FROM friendships WHERE user_a = ? AND user_b = ?').run(a, b);
  
  const ws = req.app.get('ws');
  
  if (existing.status === 'pending') {
    ws?.notify(other, { type: 'friend_request_declined', by: req.user.uuid });
  } else {
    ws?.notify(other, { type: 'friend_removed', by: req.user.uuid });
  }
  
  res.json({ ok: true });
});

router.post('/block', (req, res) => {
  const me = req.user.uuid;
  const { uuid } = req.body || {};
  if (!uuid || uuid === me) return res.status(400).json({ error: 'invalid' });
  const [a, b] = pair(me, uuid);
  db.prepare(`
    INSERT INTO friendships (user_a, user_b, status, requested_by, created_at)
    VALUES (?, ?, 'blocked', ?, ?)
    ON CONFLICT(user_a, user_b) DO UPDATE SET status = 'blocked', requested_by = excluded.requested_by
  `).run(a, b, me, Date.now());
  res.json({ ok: true });
});

module.exports = router;

File: server/friends.go
package main

import (
	"database/sql"
	"net/http"
	"strings"
)

type friendOut struct {
	UUID     string `json:"uuid"`
	Name     string `json:"name"`
	Status   string `json:"status"`
	LastSeen int64  `json:"lastSeen"`
	Since    int64  `json:"since"`
}

func handleFriendsList(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r).UUID

	rows, err := db.Query(`
		SELECT f.user_a, f.user_b, f.created_at,
		       ua.username, ua.status, ua.last_seen,
		       ub.username, ub.status, ub.last_seen
		FROM friendships f
		JOIN users ua ON ua.uuid = f.user_a
		JOIN users ub ON ub.uuid = f.user_b
		WHERE (f.user_a = ? OR f.user_b = ?) AND f.status = 'accepted'
	`, me, me)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}
	defer rows.Close()

	friends := []friendOut{}
	for rows.Next() {
		var userA, userB, usernameA, statusA, usernameB, statusB string
		var createdAt, seenA, seenB int64
		if err := rows.Scan(&userA, &userB, &createdAt, &usernameA, &statusA, &seenA, &usernameB, &statusB, &seenB); err != nil {
			continue
		}
		isA := userA == me
		f := friendOut{Since: createdAt}
		if isA {
			f.UUID, f.Name, f.Status, f.LastSeen = userB, usernameB, statusB, seenB
		} else {
			f.UUID, f.Name, f.Status, f.LastSeen = userA, usernameA, statusA, seenA
		}
		friends = append(friends, f)
	}

	writeJSON(w, http.StatusOK, M{"friends": friends})
}

type reqOut struct {
	UUID      string `json:"uuid"`
	Username  string `json:"username"`
	CreatedAt int64  `json:"created_at"`
}

func handleFriendRequests(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r).UUID

	incoming := []reqOut{}
	rows, err := db.Query(`
		SELECT f.requested_by, u.username, f.created_at
		FROM friendships f
		JOIN users u ON u.uuid = f.requested_by
		WHERE (f.user_a = ? OR f.user_b = ?) AND f.status = 'pending' AND f.requested_by != ?
	`, me, me, me)
	if err == nil {
		defer rows.Close()
		for rows.Next() {
			var o reqOut
			if rows.Scan(&o.UUID, &o.Username, &o.CreatedAt) == nil {
				incoming = append(incoming, o)
			}
		}
	}

	outgoing := []reqOut{}
	rows2, err := db.Query(`
		SELECT
			CASE WHEN f.requested_by = ? THEN f.user_b ELSE f.user_a END,
			CASE WHEN f.requested_by = ? THEN ub.username ELSE ua.username END,
			f.created_at
		FROM friendships f
		JOIN users ua ON ua.uuid = f.user_a
		JOIN users ub ON ub.uuid = f.user_b
		WHERE (f.user_a = ? OR f.user_b = ?) AND f.status = 'pending' AND f.requested_by = ?
	`, me, me, me, me, me)
	if err == nil {
		defer rows2.Close()
		for rows2.Next() {
			var o reqOut
			if rows2.Scan(&o.UUID, &o.Username, &o.CreatedAt) == nil {
				outgoing = append(outgoing, o)
			}
		}
	}

	writeJSON(w, http.StatusOK, M{"incoming": incoming, "outgoing": outgoing})
}

type friendship struct {
	ID          int64
	UserA       string
	UserB       string
	Status      string
	RequestedBy string
	CreatedAt   int64
}

func getFriendship(a, b string) (*friendship, error) {
	x, y := pair(a, b)
	var f friendship
	err := db.QueryRow(`SELECT id, user_a, user_b, status, requested_by, created_at FROM friendships WHERE user_a = ? AND user_b = ?`, x, y).
		Scan(&f.ID, &f.UserA, &f.UserB, &f.Status, &f.RequestedBy, &f.CreatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &f, nil
}

func getUserBasic(uuid string) (username, status string, ok bool) {
	err := db.QueryRow(`SELECT username, status FROM users WHERE uuid = ?`, uuid).Scan(&username, &status)
	return username, status, err == nil
}

func handleFriendAdd(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r)
	var body struct {
		Username string `json:"username"`
	}
	if err := readJSON(r, &body); err != nil || body.Username == "" || len(body.Username) > 16 {
		writeJSON(w, http.StatusBadRequest, M{"error": "invalid_username"})
		return
	}

	var targetUUID, targetName string
	err := db.QueryRow(`SELECT uuid, username FROM users WHERE username = ? COLLATE NOCASE`, body.Username).
		Scan(&targetUUID, &targetName)
	if err == sql.ErrNoRows {
		writeJSON(w, http.StatusNotFound, M{"error": "user_not_found", "hint": "L'utilisateur doit s'être connecté au moins une fois"})
		return
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}
	if targetUUID == me.UUID {
		writeJSON(w, http.StatusBadRequest, M{"error": "cannot_add_self"})
		return
	}

	a, b := pair(me.UUID, targetUUID)
	existing, err := getFriendship(me.UUID, targetUUID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}

	if existing != nil {
		switch {
		case existing.Status == "accepted":
			writeJSON(w, http.StatusConflict, M{"error": "already_friends"})
			return
		case existing.Status == "blocked":
			writeJSON(w, http.StatusForbidden, M{"error": "blocked"})
			return
		case existing.RequestedBy == me.UUID:
			writeJSON(w, http.StatusConflict, M{"error": "already_requested"})
			return
		}

		if _, err := db.Exec(`UPDATE friendships SET status = 'accepted' WHERE id = ?`, existing.ID); err != nil {
			writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
			return
		}

		hub.notify(targetUUID, M{"type": "friend_accepted", "by": me})
		if hub.isOnline(me.UUID) {
			if uname, status, ok := getUserBasic(me.UUID); ok {
				hub.notify(targetUUID, M{"type": "friend_status", "uuid": me.UUID, "name": uname, "status": status, "lastSeen": nowMillis()})
			}
		}
		if hub.isOnline(targetUUID) {
			if uname, status, ok := getUserBasic(targetUUID); ok {
				hub.notify(me.UUID, M{"type": "friend_status", "uuid": targetUUID, "name": uname, "status": status, "lastSeen": nowMillis()})
			}
		}

		writeJSON(w, http.StatusOK, M{"ok": true, "status": "accepted"})
		return
	}

	if _, err := db.Exec(`
		INSERT INTO friendships (user_a, user_b, status, requested_by, created_at)
		VALUES (?, ?, 'pending', ?, ?)
	`, a, b, me.UUID, nowMillis()); err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}

	hub.notify(targetUUID, M{"type": "friend_request", "from": me})
	writeJSON(w, http.StatusOK, M{"ok": true, "status": "pending"})
}

func handleFriendAccept(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r)
	var body struct {
		UUID string `json:"uuid"`
	}
	if err := readJSON(r, &body); err != nil || body.UUID == "" {
		writeJSON(w, http.StatusBadRequest, M{"error": "missing_uuid"})
		return
	}

	f, err := getFriendship(me.UUID, body.UUID)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}
	if f == nil || f.Status != "pending" {
		writeJSON(w, http.StatusNotFound, M{"error": "no_request"})
		return
	}
	if f.RequestedBy == me.UUID {
		writeJSON(w, http.StatusBadRequest, M{"error": "cannot_accept_own"})
		return
	}

	if _, err := db.Exec(`UPDATE friendships SET status = 'accepted' WHERE id = ?`, f.ID); err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}

	hub.notify(body.UUID, M{"type": "friend_accepted", "by": me})
	if hub.isOnline(me.UUID) {
		if uname, status, ok := getUserBasic(me.UUID); ok {
			hub.notify(body.UUID, M{"type": "friend_status", "uuid": me.UUID, "name": uname, "status": status, "lastSeen": nowMillis()})
		}
	}
	if hub.isOnline(body.UUID) {
		if uname, status, ok := getUserBasic(body.UUID); ok {
			hub.notify(me.UUID, M{"type": "friend_status", "uuid": body.UUID, "name": uname, "status": status, "lastSeen": nowMillis()})
		}
	}

	writeJSON(w, http.StatusOK, M{"ok": true})
}

func handleFriendDelete(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r)
	other := strings.TrimPrefix(r.URL.Path, "/api/friends/")
	if other == "" {
		writeJSON(w, http.StatusNotFound, M{"error": "not_found"})
		return
	}

	existing, err := getFriendship(me.UUID, other)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}
	if existing == nil {
		writeJSON(w, http.StatusNotFound, M{"error": "not_found"})
		return
	}

	a, b := pair(me.UUID, other)
	if _, err := db.Exec(`DELETE FROM friendships WHERE user_a = ? AND user_b = ?`, a, b); err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}

	if existing.Status == "pending" {
		hub.notify(other, M{"type": "friend_request_declined", "by": me.UUID})
	} else {
		hub.notify(other, M{"type": "friend_removed", "by": me.UUID})
	}

	writeJSON(w, http.StatusOK, M{"ok": true})
}

func handleFriendBlock(w http.ResponseWriter, r *http.Request) {
	me := userFromCtx(r)
	var body struct {
		UUID string `json:"uuid"`
	}
	if err := readJSON(r, &body); err != nil || body.UUID == "" || body.UUID == me.UUID {
		writeJSON(w, http.StatusBadRequest, M{"error": "invalid"})
		return
	}

	a, b := pair(me.UUID, body.UUID)
	_, err := db.Exec(`
		INSERT INTO friendships (user_a, user_b, status, requested_by, created_at)
		VALUES (?, ?, 'blocked', ?, ?)
		ON CONFLICT(user_a, user_b) DO UPDATE SET status = 'blocked', requested_by = excluded.requested_by
	`, a, b, me.UUID, nowMillis())
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, M{"error": "internal"})
		return
	}

	writeJSON(w, http.StatusOK, M{"ok": true})
}
